### PR TITLE
docs: Developer guideline to write unit/integration tests in CAPA repo

### LINKED ISSUE #3288: Update developer guide for clear distinction between unit and integration tests
/kind documentation

**Describe the solution you'd like**
As part of https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2910, we have added few integration and unit tests in controllers. We have separated two files *_unit_test.go specifically for unit tests and *_test.go for integration tests.
We need to make this clear in our developer docs.

### PR DESCRIPTION
**Pull request by Ankitasw**
**What type of PR is this?**
/kind documentation

**What this PR does / why we need it**:
As part of https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2910, we have added few integration and unit tests in controllers. We have separated two files *_unit_test.go specifically for unit tests and *_test.go for integration tests.
This PR makes clear distinction between the two files and  convention of how to differentiate integration vs unit tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3288 

**Checklist**:
- [x] squashed commits
- [X] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Updated the test conventions used for CAPA controllers.
```

### CODE REVIEW

**Comment by sedefsavas** [comment]
/approve

cc @pydctw if there is any comment.

**Comment by Ankitasw** [comment]
@sedefsavas @pydctw please let me know if you have any comments here.

File: docs/book/src/development/conventions.md [hunk 1]
```diff
@@ -36,3 +36,21 @@ use
...
+There are two types of tests written for CAPA controllers in this repo:
+* Unit tests  
+* Integration tests
+
+If any new unit or integration tests has to be added in this repo,we should follow the below conventions.
+
+### Unit tests
+These tests are written to test the functions inside the same controller file where we perform sanity checks, functionality checks etc. Here we use all types of mocking like [fakeclient](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/fake) to mock the services API calls. 
+
+These tests go into the file with suffix *_unit_test.go.
```
Comment (richardcase) [inline docs/book/src/development/conventions.md:51]: I don't think we have the test file names the right way around. Personally, `_test.go` should be unit tests and some other pattern for integration tests.

Comment (Ankitasw) [reply docs/book/src/development/conventions.md:51]: That's what we have been following from start, but I will let @sedefsavas weigh in here, if renaming has to be done, I will raise another PR for the same, and change these suffix here.

File: docs/book/src/development/conventions.md [hunk 2]
```diff
@@ -36,3 +36,21 @@ use
 ```
 
 And then within the code you can check the length or range over the slice.
+
+## Tests
+
+There are two types of tests written for CAPA controllers in this repo:
```
[the remark this answers is not in the document]
Comment (Ankitasw) [reply docs/book/src/development/conventions.md:42]: Shall I explain about E2E tests as well here, coz E2E exists in [here](https://cluster-api-aws.sigs.k8s.io/development/e2e.html).
@richardcase Probably what I can do is refer this doc here, wdyt?

[the remark this answers is not in the document]
Comment (richardcase) [reply docs/book/src/development/conventions.md:42]: Yes a reference would be good :+1:

File: docs/book/src/development/conventions.md (hunk 1 above)
Comment (richardcase) [reply docs/book/src/development/conventions.md:51]: Lets chat in office hours :) We can resolve this until then

File: docs/book/src/development/conventions.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ankitasw) [reply docs/book/src/development/conventions.md:42]: Done

File: docs/book/src/development/conventions.md [hunk 3]
```diff
@@ -36,3 +36,25 @@ use
...
+* Unit tests  
+* Integration tests
+* E2E tests
+
+If any new unit or integration tests has to be added in this repo,we should follow the below conventions.
+
+### Unit tests
+These tests are meant to verify the functions inside the same controller file where we perform sanity checks, functionality checks etc. Here we use all types of mocking like [fakeclient](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/fake) to mock the services API calls. 
+
```
Comment (pydctw) [inline docs/book/src/development/conventions.md:51]: fakeclient is not used to mock services API calls but gomock is.

File: docs/book/src/development/conventions.md [hunk 4]
```diff
@@ -36,3 +36,25 @@ use
...
+There are two types of tests written for CAPA controllers in this repo:
+* Unit tests  
+* Integration tests
+* E2E tests
+
+If any new unit or integration tests has to be added in this repo,we should follow the below conventions.
+
+### Unit tests
+These tests are meant to verify the functions inside the same controller file where we perform sanity checks, functionality checks etc. Here we use all types of mocking like [fakeclient](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client/fake) to mock the services API calls.
```
Comment (pydctw) [inline docs/book/src/development/conventions.md:50]: Both fakeclient and testEnv are used in unit tests, for example, in https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/controllers/awscluster_controller_unit_test.go. So we cannot say use fakeclient for unit test and testEnv for integration test. The choice depends on test cases as discussed in previous PRs.

Comment (Ankitasw) [reply docs/book/src/development/conventions.md:50]: Thanks @pydctw for comments, but what do you suggest, shall we not mention about the fakeclient and envtest here?

Comment (Ankitasw) [reply docs/book/src/development/conventions.md]: @pydctw wdyt now?

File: docs/book/src/development/conventions.md [hunk 5]
```diff
@@ -36,3 +36,25 @@ use
 ```
 
 And then within the code you can check the length or range over the slice.
+
+## Tests
+
+There are two types of tests written for CAPA controllers in this repo:
```
Comment (pydctw) [inline docs/book/src/development/conventions.md:42]: three types as you added E2E tests

File: docs/book/src/development/conventions.md (hunk 4 above)
Comment (Ankitasw) [reply docs/book/src/development/conventions.md]: Thankyou, if you dont mind can you lgtm it?

File: docs/book/src/development/conventions.md (hunk 5 above)
Comment (Ankitasw) [reply docs/book/src/development/conventions.md]: Done

**Pull request merged**